test/jdk/tools/jpackage/helpers/jdk/jpackage/test/PackageTest.java

```diff
@@ -393,164 +456,200 @@ protected void runAction(Action... action) {
+                .sorted(Comparator.comparing(Map.Entry::getKey))
+                .map(entry -> {
+                    return  createPackageTypeHandler(entry.getKey(), entry.getValue());
+                }).toList();
     }
 
-    private Consumer<Action> createPackageTypeHandler(
-            PackageType type, Handler handler) {
-        return toConsumer(new ThrowingConsumer<Action>() {
-            @Override
-            public void accept(Action action) throws Throwable {
-                if (terminated) {
-                    throw new IllegalStateException();
-                }
+    private record PackageTypePipeline(PackageType type, int expectedJPackageExitCode,
+            int expectedInstallExitCode, PackageHandlers packageHandlers, Handler handler,
+            JPackageCommand cmd, State state) implements Consumer<Action> {
 
-                if (action == Action.FINALIZE) {
-                    if (unpackDir != null) {
-                        if (Files.isDirectory(unpackDir)
-                                && !unpackDir.startsWith(TKit.workDir())) {
-                            TKit.deleteDirectoryRecursive(unpackDir);
-                        }
-                        unpackDir = null;
-                    }
-                    terminated = true;
-                }
+        PackageTypePipeline {
+            Objects.requireNonNull(type);
+            Objects.requireNonNull(packageHandlers);
+            Objects.requireNonNull(handler);
+            Objects.requireNonNull(cmd);
+            Objects.requireNonNull(state);
+        }
 
-                boolean skip = false;
-
-                if (unhandledAction != null) {
-                    switch (unhandledAction) {
-                        case CREATE:
-                            skip = true;
-                            break;
-                        case UNPACK:
-                        case INSTALL:
-                            skip = (action == Action.VERIFY_INSTALL);
-                            break;
-                        case UNINSTALL:
-                            skip = (action == Action.VERIFY_UNINSTALL);
-                            break;
-                        default: // NOP
-                    }
-                }
+        PackageTypePipeline(PackageType type, int expectedJPackageExitCode,
+                int expectedInstallExitCode, PackageHandlers packageHandlers,
+                Handler handler, JPackageCommand cmd) {
+            this(type, expectedJPackageExitCode, expectedInstallExitCode,
+                    packageHandlers, handler, cmd, new State());
+        }
 
-                if (skip) {
+        @Override
+        public void accept(Action action) {
+            switch(analizeAction(action)) {
+                case SKIP_NO_PACKAGE_HANDLER -> {
+                    TKit.trace(String.format("No handler of [%s] action for %s command",
+                            action, cmd.getPrintableCommandLine()));
+                    return;
```

Comment:
Do we really need `return` here and at line 495? Also, do you think it might be better to make `switch` consistent? For example this one uses `->` and below `:`.

Reply:
> Do we really need return here and at line 495?

Yes. It should exit the function early to avoid jumping in the next `switch` block.

> do you think it might be better to make `switch` consistent?

Totally!

Reply:
> Yes. It should exit the function early to avoid jumping in the next switch block.
With `->` it should not fall into next case. Maybe I am missing something.

Reply:
> With `->` it should not fall into next case. Maybe I am missing something.

Right, but there are two `siwtch` blocks in the function. Without `return` the control flow will get to the second one. `returns` prevents this.